File: M_Simulator_FEM_2D/08_M_Simulator_FEM_2D/Python/src/Build_Force.py
import numpy as np

def Build_Force(Gen):


    # =============================================================================
    # Find where to apply force
    # =============================================================================

    left_nodes = np.arange(0, Gen["Nn"], Gen["Nx"] + 1)
    right_nodes = np.arange(Gen["Nx"], Gen["Nn"], Gen["Nx"] + 1)
    bottom_nodes = np.arange(0, Gen["Nx"] + 1)
    top_nodes = np.arange(Gen["Nn"] - Gen["Nx"] - 1, Gen["Nn"])

    # DOFs
    left_x = 2 * left_nodes
    right_x = 2 * right_nodes
    bottom_y = 2 * bottom_nodes + 1
    top_y = 2 * top_nodes + 1


    # =============================================================================
    # Make force vector
    # =============================================================================

    F = np.zeros(2*Gen["Nn"])

    #Forces to be applied (per unit of out of plane thickness) [N/m]
    fy = -10e7
    fx = -8.5e7

    #To which nodes to we apply it
    #On top nodes
    F[top_y[1:-1]] = fy
    #On right nodes
    F[right_x[1:-1]] = fx
    #On bottom nodes
    # F[bottom_y[1:-1]] = -fy
    #On left nodes
    # F[left_x[1:-1]] = -fx

    #Only apply half the force on the ends
    #Top left corner
    F[top_y[0]] = fy / 2
    #Top right corner
    F[top_y[-1]] = fy / 2

    #Right bottom corner
    F[right_x[0]] = fx/2
    #Right top corner
    F[right_x[-1]] = fx/2

    #Bottom left corner
    # F[bottom_y[0]] = -fy/2
    #Bottom right corner
    # F[bottom_y[-1]] = -fy/2

    #Left bottom corner
    # F[left_x[0]] = -fx/2
    #Left top corner
    # F[left_x[-1]] = -fx/2

    return F

File: HM_Simulator_FEM_FVM_2D/09_HM_Simulator_FEM_FVM_2D/MATLAB/src/Mechanics_Model/Build_Force.m
function [F] = Build_Force(Gen)

%Where to apply the force (on all top nodes)
A = (2*Gen.Nn-Gen.Nx*2+2):2:2*Gen.Nn-2;       
%On all right nodes
B = 4*(Gen.Nx+1)-1:2*(Gen.Nx+1):2*Gen.Nn-2*(Gen.Nx+1);
%On all bottom nodes
C = 4:2:2*Gen.Nx;
%On all left nodes
D = 2*(Gen.Nx+1)+1:2*(Gen.Nx+1):2*Gen.Nn-4*(Gen.Nx+1)+1;
%Set up force vector
F = zeros(2*Gen.Nn,1);                                      %[2*Nn,1]

%Apply force (per unit of out of plane thickness) [N/m]
fy = -0e7;
fx = -0e7;

%On top nodes
F(A,1) = fy;
%On right nodes
F(B,1) = fx;
%On bottom nodes
% F(C,1) = -fy;
%On left nodes
% F(D,1) = -fx;

%End of force 
%On top left node y direction
F(2*Gen.Nn-Gen.Nx*2,1) = fy/2;
%On top left node x direction
% F(2*Gen.Nn-Gen.Nx*2-1,1) = -fx/2;
%On top right node y direction
F(2*Gen.Nn,1) = fy/2;
%On top right node x direction
F(2*Gen.Nn-1,1) = fx/2;
%On bottom right node y dirction
% F(2*(Gen.Nx+1),1) = -fy/2;
%On bottom right node x direction
F(2*(Gen.Nx+1)-1,1) = fx/2;
%On bottom left node y direction
% F(2,1) = -fy/2;
%On bottom left node x direction
% F(1,1) = -fx/2;

%Add pressure contributions to force vector
% [F] = Add_Pressure(F,Gen,Res.P);                            %[2*Nn,1]

end